service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/update/UpdateControlledBigQueryDatasetsLifetimeStep.java

```diff
@@ -0,0 +1,101 @@
+    this.resourceDao = resourceDao;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    var workingMap = context.getWorkingMap();
+    validateRequiredEntries(
+        workingMap,
+        CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_TABLE_LIFETIME_MAP,
+        CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_PARTITION_LIFETIME_MAP,
+        CONTROLLED_RESOURCE_ID_TO_WORKSPACE_ID_MAP);
+
+    Map<UUID, String> resourceIdToDefaultTableLifetimeMap =
+        Preconditions.checkNotNull(
+            workingMap.get(
+                CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_TABLE_LIFETIME_MAP,
+                new TypeReference<>() {}));
+    Map<UUID, String> resourceIdToDefaultPartitionLifetimeMap =
+        Preconditions.checkNotNull(
+            workingMap.get(
+                CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_PARTITION_LIFETIME_MAP,
+                new TypeReference<>() {}));
+
+    Map<UUID, String> resourceIdsToWorkspaceIdMap =
+        Preconditions.checkNotNull(
+            workingMap.get(CONTROLLED_RESOURCE_ID_TO_WORKSPACE_ID_MAP, new TypeReference<>() {}));
+    List<ControlledBigQueryDatasetResource> updatedResources = new ArrayList<>();
+
+    List<ControlledBigQueryDatasetResource> controlledBigQueryDatasets =
+        Preconditions.checkNotNull(
+            workingMap.get(
+                CONTROLLED_BIG_QUERY_DATASETS_WITHOUT_LIFETIME, new TypeReference<>() {}));
+
+    for (var resource : controlledBigQueryDatasets) {
+      var id = resource.getResourceId();
+      if (resourceIdsToWorkspaceIdMap.containsKey(id)) {
+        boolean updated =
+            resourceDao.updateBigQueryDatasetDefaultTableAndPartitionLifetime(
+                resource.castByEnum(CONTROLLED_GCP_BIG_QUERY_DATASET),
+                Long.valueOf(resourceIdToDefaultTableLifetimeMap.get(id)),
+                Long.valueOf(resourceIdToDefaultPartitionLifetimeMap.get(id)));
+
+        if (updated) {
+          updatedResources.add(
+              resourceDao
+                  .getResource(UUID.fromString(resourceIdsToWorkspaceIdMap.get(id)), id)
+                  .castToControlledResource()
+                  .castByEnum(CONTROLLED_GCP_BIG_QUERY_DATASET));
+        }
+      }
+    }
+
+    setResponse(context, updatedResources, HttpStatus.OK);
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    List<ControlledBigQueryDatasetResource> controlledBigQueryDatasets =
+        context
+            .getWorkingMap()
+            .get(CONTROLLED_BIG_QUERY_DATASETS_WITHOUT_LIFETIME, new TypeReference<>() {});
+
+    if (controlledBigQueryDatasets == null || controlledBigQueryDatasets.isEmpty()) {
+      return StepResult.getStepResultSuccess();
+    }
+    for (var resource : controlledBigQueryDatasets) {
+      resourceDao.updateBigQueryDatasetDefaultTableAndPartitionLifetime(
```

Comment:
I'm very confused by this code. In what way does this undo actually undo anything?
Won't it be filling in things in the attributes that were not there before? Are the values correct? Are they null? Where did they come from?

Reply:
the resourceDao logic to retrieve bq dataset without lifetime uses OR so the resources may have one of the attributes set. 

I think it's probably enough to use AND because the old resource definitely don't have either but this is fine with me too..